apstra/talk_to_api_ops.go

```diff
@@ -0,0 +1,359 @@
+	switch {
+	case !gzResponse && !s3Response:
+		innerResp.Body = io.NopCloser(base64.NewDecoder(base64.StdEncoding, strings.NewReader(proxyResponse.Response)))
+
+	case !gzResponse && s3Response:
+		s3Resp, err := getUrl(ctx, responseS3Url, o.httpClient)
+		if err != nil {
+			return fmt.Errorf("fetching response from S3 URL %q: %w", responseS3Url, err)
+		}
+
+		innerResp.Body = s3Resp.Body
+
+	case gzResponse && !s3Response:
+		gzReader, err := gzip.NewReader(base64.NewDecoder(base64.StdEncoding, strings.NewReader(proxyResponse.Response)))
+		if err != nil {
+			return fmt.Errorf("error creating gzip reader for transaction %s - %w", proxyResponse.Uid, err)
+		}
+
+		innerResp.Body = gzReader
+
+	case gzResponse && s3Response:
+		s3Resp, err := getUrl(ctx, responseS3Url, o.httpClient)
+		if err != nil {
+			return fmt.Errorf("fetching response from S3 URL %q: %w", responseS3Url, err)
+		}
+
+		r, err := gzip.NewReader(s3Resp.Body)
+		if err != nil {
+			return fmt.Errorf("creating gz reader for S3 response body: %w", err)
+		}
+
+		defer func() { _ = s3Resp.Body.Close() }()
+
+		innerResp.Body = r
+	}
+
+	if innerResp.StatusCode/100 != 2 {
+		req.URL = apstraUrl
+		req.URL.Host = "api-ops"
+		return newTalkToApstraErr(req, msg.Body, innerResp, "")
+	}
+
+	// If the caller gave us an httpBodyWriter, copy the response body into it and return
+	if in.httpBodyWriter != nil {
+		_, err = io.CopyBuffer(in.httpBodyWriter, innerResp.Body, nil)
+		if err != nil {
+			return fmt.Errorf("error while reading http response body - %w", err)
+		}
+		return nil
+	}
+
+	// figure out whether Apstra responded with a task ID
+	var tIdR taskIdResponse
+	taskResponseFound, err := peekParseResponseBodyAsTaskId(innerResp, &tIdR)
+	if err != nil {
+		req.URL = apstraUrl
+		req.URL.Host = "api-ops"
+		return newTalkToApstraErr(req, requestBody, innerResp, "error peeking response body")
+	}
+
+	// no task ID response, so no polling tomfoolery required
+	if !taskResponseFound {
+		if in.apiResponse == nil {
+			o.Log(2, "no task ID response, and caller wants nothing back - talkToApstra done")
+			// caller expects no response, so we're done here
+			return nil
+		}
+		o.Log(2, "no task ID response, parse apstra reply for caller")
+		// no task ID, decode response body into the caller-specified structure
+		return json.NewDecoder(innerResp.Body).Decode(in.apiResponse)
+	}
+
+	// we got a task ID, instead of the expected response object. tasks are
+	// per-blueprint, so we need to figure out the blueprint ID for task
+	// progress query reasons.
+	var bpId ObjectId
+
+	// maybe the blueprintId is in the URL?
+	if strings.Contains(apstraUrl.Path, apiUrlBlueprintsPrefix) {
+		bpId = blueprintIdFromUrl(apstraUrl)
+	}
+
+	switch {
+	case (bpId != "" && tIdR.BlueprintId != "") && (bpId != tIdR.BlueprintId):
+		return fmt.Errorf("blueprint Id in URL ('%s') and returned object body ('%s') don't match", bpId, tIdR.BlueprintId)
+	case bpId == "" && tIdR.BlueprintId == "":
```

Comment:
Is there an assumption that there will always be a blueprint targeted?
What happens if we need to create/update a property set?

Reply:
I had not seen that  this is in the task response section.